hack/get-terraform.sh

```diff
@@ -24,8 +24,15 @@ then
 fi &&
 TERRAFORM_VERSION="0.11.8" &&
 TERRAFORM_URL="https://releases.hashicorp.com/terraform/${TERRAFORM_VERSION}/terraform_${TERRAFORM_VERSION}_${OS}_${ARCH}.zip" &&
-echo "pulling ${TERRAFORM_URL}" >&2 &&
+TERRAFORM_BINARY=".cache/terraform_${TERRAFORM_VERSION}_${OS}_${ARCH}"
 cd "$(dirname "$0")/.." &&
-mkdir -p bin &&
-curl -L "${TERRAFORM_URL}" | "${FUNZIP}" >bin/terraform &&
-chmod +x bin/terraform
+if [ ! -x "${TERRAFORM_BINARY}" ]
+then
+	echo "pulling ${TERRAFORM_URL}" >&2 &&
+	mkdir -p "$(dirname "${TERRAFORM_BINARY}")" &&
+	curl -L "${TERRAFORM_URL}" | "${FUNZIP}" >"${TERRAFORM_BINARY}" &&
+	chmod +x "${TERRAFORM_BINARY}"
+fi &&
+mkdir -p data/data/bin &&
+rm -f data/data/bin/terraform &&
+ln -s "../../../${TERRAFORM_BINARY}" data/data/bin/terraform
```

Comment:
@wking do you know a good way of calculating `"../../.."` from `"data/data/bin/terraform"`?

Reply:
I came up with the following, but it may be more effort than it is worth:

```diff
diff --git a/hack/get-terraform.sh b/hack/get-terraform.sh
index aae1b69cd..c47aaa11e 100755
--- a/hack/get-terraform.sh
+++ b/hack/get-terraform.sh
@@ -25,6 +25,7 @@ fi &&
 TERRAFORM_VERSION="0.11.8" &&
 TERRAFORM_URL="https://releases.hashicorp.com/terraform/${TERRAFORM_VERSION}/terraform_${TERRAFORM_VERSION}_${OS}_${ARCH}.zip" &&
 TERRAFORM_BINARY=".cache/terraform_${TERRAFORM_VERSION}_${OS}_${ARCH}"
+TERRAFORM_SYMLINK="data/data/bin/terraform"
 cd "$(dirname "$0")/.." &&
 if [ ! -x "${TERRAFORM_BINARY}" ]
 then
@@ -33,6 +34,6 @@ then
        curl -L "${TERRAFORM_URL}" | "${FUNZIP}" >"${TERRAFORM_BINARY}" &&
        chmod +x "${TERRAFORM_BINARY}"
 fi &&
-mkdir -p data/data/bin &&
-rm -f data/data/bin/terraform &&
-ln -s "../../../${TERRAFORM_BINARY}" data/data/bin/terraform
+mkdir -p "$(dirname "${TERRAFORM_SYMLINK}")" &&
+rm -f "${TERRAFORM_SYMLINK}" &&
+ln -s "$(dirname "${TERRAFORM_SYMLINK}" | sed 's_[^/]*_.._g')/${TERRAFORM_BINARY}" "${TERRAFORM_SYMLINK}"
```

Reply:
> do you know a good way of calculating `"../../.."` from `"data/data/bin/terraform"`?

Well, one way would be to use hardlinks instead of symlinks:

```console
$ rm -f data/data/bin/terraform
$ ls -l .cache/terraform_0.11.8_linux_amd64 
-rwxr-xr-x. 1 trking trking 76422048 Dec  6 21:54 .cache/terraform_0.11.8_linux_amd64
$ ln .cache/terraform_0.11.8_linux_amd64 data/data/bin/terraform
$ ls -l .cache/terraform_0.11.8_linux_amd64 
-rwxr-xr-x. 2 trking trking 76422048 Dec  6 21:54 .cache/terraform_0.11.8_linux_amd64
$ ls -l data/data/bin/terraform 
-rwxr-xr-x. 2 trking trking 76422048 Dec  6 21:54 data/data/bin/terraform
```

See the [`<number of links>`][1] field changing from 1 to 2 after the `ln` call.

[1]: http://pubs.opengroup.org/onlinepubs/9699919799/utilities/ls.html#tag_20_73_10

Reply:
I assume all of this will always be on the same filesystem, so a hardlink should work. I'll go with that instead.